syncer/syncer.go

```diff
@@ -718,6 +731,68 @@ func (s *Syncer) addCount(isFinished bool, queueBucket string, tp opType, n int6
 	}
 }
 
+func (s *Syncer) jobsIsEmpty() bool {
+	for idx := range s.jobs {
+		if len(s.jobs[idx]) > 0 {
+			return false
+		}
+	}
+	return true
+}
+
+// updateReplicationLag calculates syncer's replication lag by job, it called after every batch dml job / one skip job / one ddl
+// job is flushed to target db.
+func (s *Syncer) updateReplicationLag(job *job, queueBucketName string) {
+	// when job is nil mean no job in this bucket, need do reset this bucket lag to 0
+	if job == nil {
+		s.workerLagMap[queueBucketName].Store(0)
+		// when job is nill and s.jobs is emepty, means all event is consumed, we update lag to 0
+		if s.jobsIsEmpty() {
+			replicationLagGauge.WithLabelValues(s.cfg.Name).Set(float64(0))
+			s.secondsBehindMaster.Store(0)
+		}
+		return
+	}
+
+	failpoint.Inject("BlockSyncerUpdateLag", func(v failpoint.Value) {
+		sv, ok := v.(string)
+		if ok && len(strings.Split(sv, ",")) == 2 {
+			args := strings.Split(sv, ",")
+			jtp := args[0]                // job type
+			t, _ := strconv.Atoi(args[1]) // sleep time
+			if job.tp.String() == jtp {
+				s.tctx.L().Info("BlockSyncerUpdateLag", zap.String("job type", jtp), zap.Int("sleep time", t))
+				time.Sleep(time.Second * time.Duration(t))
+			}
+		} else {
+			s.tctx.L().Error("BlockSyncerUpdateLag failed", zap.Any("input v", v))
+			panic("BlockSyncerUpdateLag failed")
+		}
+	})
+
+	var lag int64
+	switch job.tp {
+	case ddl:
+		// NOTE: we handle ddl job separately because ddl job will clean all the dml job before execution
+		lag = time.Now().Unix() - s.tsOffset.Load() - int64(job.eventHeader.Timestamp)
+	default: // dml job or skip job
+		if job.tp != skip {
+			// NOTE workerlagmap already init all dml key(queueBucketName) before syncer running.
+			s.workerLagMap[queueBucketName].Store(time.Now().Unix() - s.tsOffset.Load() - int64(job.eventHeader.Timestamp))
+		} else {
+			lag = time.Now().Unix() - s.tsOffset.Load() - int64(job.eventHeader.Timestamp)
+		}
+		// find all job queue lag choose the max one
```

Comment:
I think this logic could handle workerLagMap all zero, so above if could be shorter.

Reply:
i think check  job queue is empty directly has more clean logic than check the `workerLagMap` ?

Reply:
I agree to lance. To me this logic can be changed.